### PR TITLE
KNIDEPLOY-4151 Operator deploys the assisted installer 

### PR DESCRIPTION
**Pull request by rwsu**
Adds --apply-manifest option to create manifests without deploying them to a cluster. This option is used by the create-ocp-manifests Makefile target. The manifests are written to build/assisted-installer.

A new "bundle" target has been added to the Makefile. This target generates the operator bundle manifests and CSV using the manifests in build/assisted-installer as the source.

main/cmd.go has been updated to include code to create a route for the service. This is only enabled when CREATE_SERVICE_ROUTE environment variable is set to true and by default is false. The SERVICE_BASE_URL is then updated using the route's URL as the value.

CREATE_SERVICE_ROUTE is set to true using a kustomize patch in config/default/assisted-service-configmap-patch-create-service-route.yaml. 

The operator currently deploys scality. A future PR will remove scality/s3 and use the filesystem implementation (https://issues.redhat.com/browse/KNIDEPLOY-4234). In the meantime, to have a working operator deployment, scality is used and it requires setting DEPLOY_TARGET to "k8s". This is enabled through a kustomize patch, config/default/assisted-service-patch-deploy-target.yaml. This is required because setting DEPLOY_TARGET to "ocp" (the eventual target) will force the s3 filesystem implementation which is not ready for use because additional work is needed to resolve a missing coreos-installer issue (https://issues.redhat.com/browse/MGMT-4072).

### CODE REVIEW

File: config/certmanager/certificate.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# The following manifests contain a self-signed issuer CR and a certificate CR.
```
Comment (asalkeld) [inline config/certmanager/certificate.yaml:1]: are you using certmanager? If not, probably remove this directrory else it's probably better to use https://docs.openshift.com/container-platform/4.6/security/certificates/service-serving-certificate.html

File: bundle.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,15 @@
+FROM scratch
```
Comment (djzager) [inline bundle.Dockerfile:1]: An alternative to building from scratch would be to use a multi-stage build and use `make` targets to get everything ready for building a bundle.

File: config/assisted-service/assisted-installer-sso.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
```
Comment (djzager) [inline config/assisted-service/assisted-installer-sso.yaml:1]: Unless we intend on moving away from all of the stuff in `deploy/`...I am hesitant to recommend including all of the kustomize stuff in the source tree.

It isn't immediately clear if we need all of the bundle manifests included for midstream/downstream, but that isn't a problem to be solved now :sunglasses: 

**tl;dr** either we include all of the manifests in the source tree (kustomize based deployment) and move away from the python deployment mechanisms or we generate the kustomize based manifests only when we need them to build a bundle.

**Review by djzager — CHANGES_REQUESTED** [review]
I think this may be a good thing for us to discuss in the architecture forum later this week, if we can -- and if we can whether or not we should -- move to kustomize for deploying the assisted-service.

There are definitely advantages to falling more in line with what operator-sdk/kubebuilder expect but I bet there are some costs to that approach.

File: PROJECT [hunk 4]
```diff
@@ -0,0 +1,13 @@
+domain: openshift.io
+layout: go.kubebuilder.io/v3
+projectName: assisted-service-operator
+repo: github.com/openshift/assisted-service-operator
```
Comment (omertuc) [inline PROJECT:4]: What is a PROJECT file? Why is it required? Does it have to have this name? Is this line correct?

File: Makefile [hunk 5]
```diff
@@ -338,3 +350,36 @@ delete-minikube-profile:
...
+############
+# Operator #
+############
+
+# Current Operator version
+OPERATOR_VERSION ?= 0.0.1
+
+# Generate bundle manifests and metadata, then validate generated files.
+.PHONY: bundle
+bundle: create-ocp-manifests
```
Comment (omertuc) [inline Makefile:363]: `operator-bundle`

File: Makefile [hunk 6]
```diff
@@ -338,3 +350,36 @@ delete-minikube-profile:
...
+	cp ./build/assisted-installer/assisted-service.yaml config/assisted-service
+	cp ./build/assisted-installer/deploy_ui.yaml config/assisted-service
+	#operator-sdk generate kustomize manifests -q
+	#cd config/manager && $(KUSTOMIZE) edit set image controller=$(IMG)
+	kustomize build config/manifests | operator-sdk generate bundle -q --overwrite --version $(OPERATOR_VERSION) $(BUNDLE_METADATA_OPTS)
+	operator-sdk bundle validate ./bundle
+
+# Build the bundle image.
+.PHONY: bundle-build
+bundle-build:
```
Comment (omertuc) [inline Makefile:384]: `operator-bundle-build`

File: config/assisted-service/postgres-secret.yaml [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+kind: Secret
+metadata:
+  labels:
+    app: postgres
+  name: assisted-installer-rds
+  namespace: assisted-installer
+stringData:
+  db.host: postgres
+  db.name: installer
+  db.password: admin
```
Comment (carbonin) [inline config/assisted-service/postgres-secret.yaml:11]: Can this be randomly generated somehow?
Or is this not the actual password that will get used?

File: config/assisted-service/scality-secret.yaml [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+metadata:
+  name: assisted-installer-s3
+  namespace: "assisted-installer"
+  labels:
+    app: scality
+type: Opaque
+stringData:
+  aws_access_key_id: "accessKey1"
+  aws_region: "us-east-1"
+  aws_secret_access_key: "verySecretKey1"
```
Comment (carbonin) [inline config/assisted-service/scality-secret.yaml:12]: Same here, can we generate this?

File: cmd/main.go [hunk 9]
```diff
@@ -555,3 +566,82 @@ func createControllerManager() (manager.Manager, error) {
 	}
 	return nil, nil
 }
+
+func newAssistedServiceRoute(service *corev1.Service) *routev1.Route {
```
Comment (carbonin) [inline cmd/main.go:570]: Nothing in here looks conditional. Can we create this outside of the app and send in the route name to do the rest of the logic?

Ideally the application wouldn't know about any of this and should just be sent the externally accessible endpoint, but if that's not possible we should get as much of this out of the app as we can.

File: config/rbac/controller_roles.yaml [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+  - ""
+  resources:
+  - secrets
+  verbs:
+  - get
+  - list
+  - update
+  - watch
+- apiGroups:
+  - adi.io.my.domain
```
Comment (carbonin) [inline config/rbac/controller_roles.yaml:19]: What is this group? Should this be something more meaningful? Or is this something that someone has already defined elsewhere?

File: config/samples/assistedservice.openshift.io_v1alpha1_assistedservice.yaml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+apiVersion: assistedservice.openshift.io.openshift.io/v1alpha1
+kind: AssistedService
+metadata:
+  name: assistedservice-sample
+spec:
+  # Add fields here
+  foo: bar
```
Comment (carbonin) [inline config/samples/assistedservice.openshift.io_v1alpha1_assistedservice.yaml:7]: Should this be a real example?

File: docs/operator.md [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: assisted-installer
+  labels:
+    name: assisted-installer
+EOF
+```
+
+Create two PVCs, one for postgres and another for s3 (temporary until filesystem implementation is available).
```
Comment (carbonin) [inline docs/operator.md:50]: Why don't we create the PVC as a part of the deployment?
Also I don't see the volume mount in the PG deployment, how does it end up using this?

File: config/assisted-service/scality-secret.yaml (hunk 8 above)
Comment (celebdor) [reply config/assisted-service/scality-secret.yaml:12]: In this case it doesn't matter as much, because a follow-up PR will remove S3 involvement

File: config/rbac/controller_roles.yaml (hunk 10 above)
Comment (celebdor) [reply config/rbac/controller_roles.yaml:19]: definitely should be more meaningful

File: PROJECT (hunk 4 above)
Comment (rwsu) [reply PROJECT:4]: The PROJECT file is read by "operator-sdk generate bundle" to determine the project name. The value for this line isn't correct. After looking at this some more, we can remove this file and just pass in the project name when calling "operator-sdk generate bundle".

File: cmd/main.go (hunk 9 above)
Comment (djzager) [reply cmd/main.go:570]: This comment does make me wonder if we could do this work as an `initContainer` instead of app startup. There isn't a way to handle it via OLM packaging mechanisms...so we would either need to make it a pre-req for the user, check out initContainers as a solution, or as part of app startup.

File: config/assisted-service/assisted-service-configmap.yaml [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+  CHECK_CLUSTER_VERSION: 'False'
+  CONTROLLER_IMAGE: quay.io/ocpmetal/assisted-installer-controller:latest
+  CREATE_S3_BUCKET: 'true'
+  ENABLE_AUTH: 'False'
+  ENABLE_KUBE_API: 'true'
+  INSTALLER_IMAGE: quay.io/ocpmetal/assisted-installer:latest
+  JWKS_URL: https://api.openshift.com/.well-known/jwks.json
+  NAMESPACE: assisted-installer
+  OCM_BASE_URL: https://api-integration.6943.hive-integration.openshiftapps.com
+  OPENSHIFT_VERSIONS: '{"4.6":{"display_name":"4.6.12","release_image":"quay.io/openshift-release-dev/ocp-release:4.6.12-x86_64","rhcos_image":"https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.6/4.6.8/rhcos-4.6.8-x86_64-live.x86_64.iso","rhcos_version":"46.82.202012051820-0","support_level":"production"},"4.7":{"display_name":"4.7-pre-release","release_image":"quay.io/openshift-release-dev/ocp-release:4.7.0-rc.0-x86_64","rhcos_image":"https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/pre-release/4.7.0-fc.4/rhcos-4.7.0-fc.4-x86_64-live.x86_64.iso","rhcos_version":"47.83.202101171239-0","support_level":"beta"}}'
```
Comment (carbonin) [inline config/assisted-service/assisted-service-configmap.yaml:15]: Is there any way we can avoid adding another place we store this kind of information?

File: cmd/main.go (hunk 9 above)
Comment (carbonin) [reply cmd/main.go:570]: > There isn't a way to handle it via OLM packaging mechanisms

Can you elaborate on this? Couldn't this be added the same way we add the service? What's the difference?

File: config/assisted-service/assisted-service-configmap.yaml (hunk 13 above)
Comment (omertuc) [reply config/assisted-service/assisted-service-configmap.yaml:15]: Please add an option to generate this out of https://github.com/openshift/assisted-service/blob/7d53cb0a8ece947fc63d15756ea2b0eb787a7370/default_ocp_versions.json

Also please verify that this generation gets done in https://github.com/eranco74/assisted-installer-deployment/blob/master/tools/new_minor_OCP_version.py

File: docs/operator.md (hunk 12 above)
Comment (djzager) [reply docs/operator.md:50]: For the first question, unfortunately PVC can't be specified in the operator's manifests https://docs.openshift.com/container-platform/4.6/operators/understanding/olm-packaging-format.html

Comment (carbonin) [reply docs/operator.md:50]: So that says:
> When these optional objects are included in a bundle, Operator Lifecycle Manager (OLM) can create them from the bundle and manage their lifecycle along with the CSV

That doesn't exactly mean that we *can't* create other objects as a part of the manifest, just that they won't follow the operator's lifecycle. Which in the case of a PVC is probably good as removing it would cause data loss.
If we do include it in the spec somewhere does some part of this process (bundling or installing, presumably) error out?

File: config/assisted-service/assisted-service-configmap.yaml (hunk 13 above)
Comment (rwsu) [reply config/assisted-service/assisted-service-configmap.yaml:15]: I'm going to remove most of the yaml in config/assisted-service. They are just copies of the manifests generated into the build/assisted-installer directory. So far I haven't discovered a need to archive them into the repo, so I'll add them .gitignore.

File: config/certmanager/certificate.yaml (hunk 1 above)
Comment (rwsu) [reply config/certmanager/certificate.yaml:1]: certmanager isn't used. I'll remove this.

File: config/rbac/controller_roles.yaml (hunk 10 above)
Comment (rwsu) [reply config/rbac/controller_roles.yaml:19]: It is being defined in https://github.com/openshift/assisted-service/blob/master/internal/controller/config/rbac/role.yaml#L19. The CRDs are also using adi.io.my.domain: https://github.com/openshift/assisted-service/tree/master/internal/controller/config/crd/bases @filanov Will the apiGroup be renamed sometime in the future?

File: config/samples/assistedservice.openshift.io_v1alpha1_assistedservice.yaml (hunk 11 above)
Comment (rwsu) [reply config/samples/assistedservice.openshift.io_v1alpha1_assistedservice.yaml:7]: No :). Will remove because the AssistedService CRD doesn't exist.

File: cmd/main.go (hunk 9 above)
Comment (djzager) [reply cmd/main.go:570]: https://docs.openshift.com/container-platform/4.6/operators/understanding/olm-packaging-format.html

Specifically, under Manifests->Additionally supported types:
> Supported optional object types
Secret
ConfigMap
Service
PodDisruptionBudget
PriorityClass
VerticalPodAutoScaler

File: docs/operator.md (hunk 12 above)
Comment (djzager) [reply docs/operator.md:50]: I'm intrigued :thinking: need to think about this.

File: config/rbac/controller_roles.yaml (hunk 10 above)
Comment (filanov) [reply config/rbac/controller_roles.yaml:19]: We need to remove the domain. 
ADI stand for agent driven installer. 
I leave the naming to @avishayt and @mhrivnak

File: bundle.Dockerfile (hunk 2 above)
Comment (filanov) [inline bundle.Dockerfile:1]: shouldn't this file be called `Dockerfile.bundle`? (all the other docker files start with Dockerfile)

File: cmd/main.go [hunk 14]
```diff
@@ -555,3 +566,82 @@ func createControllerManager() (manager.Manager, error) {
...
+	failOnError(routev1.AddToScheme(scheme.Scheme),
+		"Failed to add routev1 to scheme")
+
+	client, err := client.New(config.GetConfigOrDie(), client.Options{Scheme: scheme.Scheme})
+	failOnError(err, "failed to create controller-runtime client")
+
+	ctx := context.Background()
+
+	serviceName := "assisted-service"
+	namespace := "assisted-installer"
```
Comment (filanov) [inline cmd/main.go:610]: Do we always deploy to this namespace or is it something optional?

File: config/assisted-service/assisted-service-configmap.yaml [hunk 15]
```diff
@@ -0,0 +1,25 @@
+apiVersion: v1
```
Comment (filanov) [inline config/assisted-service/assisted-service-configmap.yaml:1]: can't we use https://github.com/openshift/assisted-service/blob/master/openshift/template.yaml?
It will be the third file defining those parameters

File: config/assisted-service/assisted-service-service.yaml [hunk 16]
```diff
@@ -0,0 +1,18 @@
+apiVersion: v1
```
Comment (filanov) [inline config/assisted-service/assisted-service-service.yaml:1]: Why do create everything again when we have https://github.com/openshift/assisted-service/tree/master/deploy?

File: config/rbac/controller_roles.yaml [hunk 17]
```diff
@@ -0,0 +1,97 @@
+
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  creationTimestamp: null
+  name: manager-role
```
Comment (filanov) [inline config/rbac/controller_roles.yaml:7]: Probably need to change it to something specific no?

File: config/rbac/leader_election_role.yaml [hunk 18]
```diff
@@ -0,0 +1,27 @@
+# permissions to do leader election.
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: leader-election-role
```
Comment (filanov) [inline config/rbac/leader_election_role.yaml:5]: Maybe need to change the name to something more specific?

File: docs/operator.md (hunk 12 above)
Comment (rwsu) [reply docs/operator.md:50]: You're right, the deployment doesn't define a volume mount, so the PVC ends up being unused. Since we are intending to support persistence, the volume mount will need to be added..

File: cmd/main.go (hunk 14 above)
Comment (djzager) [reply cmd/main.go:610]: excellent point. This should probably come using the downward API.

File: config/manifests/bases/assisted-service-operator.clusterserviceversion.yaml [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+    description: Assisted Service is used to orchestrate baremetal OpenShift installations.
+    operators.operatorframework.io/builder: operator-sdk-v1.3.0
+    operators.operatorframework.io/project_layout: go.kubebuilder.io/v3
+    "operatorframework.io/suggested-namespace": assisted-installer
+  name: assisted-service-operator.v0.0.1
+spec:
+  apiservicedefinitions: {}
+  customresourcedefinitions:
+    owned:
+    - kind: Host
```
Comment (filanov) [inline config/manifests/bases/assisted-service-operator.clusterserviceversion.yaml:16]: The names have changes and not everything here is relevant. 
Need to deploy:
Agents.adi.io.my.domain
InstallEnv.adi.io.my.domain
Cluster deployment should come with hive.
Cluster and Image are not relevant anymore.

**Comment by filanov** [comment]
Just making sure that we have permissions to edit Hive cluster deployment.

**Comment by filanov** [comment]
In addition we don't need to deploy UI, with hive and kube-api we expect the users not to use rest-api, deploying the UI may be confusing.

**Comment by rwsu** [comment]
> In addition we don't need to deploy UI, with hive and kube-api we expect the users not to use rest-api, deploying the UI may be confusing.

If the users will not use the rest-api, does this mean we don't need to create route for the service?

File: config/assisted-service/assisted-service-configmap.yaml (hunk 15 above)
Comment (rwsu) [reply config/assisted-service/assisted-service-configmap.yaml]: I wasn't sure if we needed to archive this file or even the files under ./bundle. This file has been removed.

File: config/assisted-service/assisted-service-service.yaml (hunk 16 above)
Comment (rwsu) [reply config/assisted-service/assisted-service-service.yaml]: This file has been removed.

File: config/assisted-service/assisted-service-configmap.yaml (hunk 13 above)
Comment (rwsu) [reply config/assisted-service/assisted-service-configmap.yaml]: File removed.

File: config/assisted-service/assisted-installer-sso.yaml (hunk 3 above)
Comment (rwsu) [reply config/assisted-service/assisted-installer-sso.yaml]: I've removed the files.

File: cmd/main.go (hunk 9 above)
Comment (rwsu) [reply cmd/main.go:570]: I just experimented with the init container approach.

`
      - name: init-add-route-and-update-config-map
        image: quay.io/ocpmetal/assisted-service:latest
        command: ['sh', '-c', 'export NAMESPACE=$(cat /var/run/secrets/kubernetes.io/serviceaccount/namespace) && oc expose service -n $NAMESPACE assisted-service && export ROUTE_URL=$(oc get routes -n $NAMESPACE assisted-service -o jsonpath={.spec.host}) && oc get configmap -n $NAMESPACE assisted-service-config -o yaml | sed -e "s|REPLACE_BASE_URL|http\:\/\/$ROUTE_URL|" | oc apply -f - ']
`

It works fine, creating the route and updating the SERVICE_BASE_URL in the config map. But it runs into the problem where the route remains when the operator is uninstalled.  Reinstalling the operator throws an error in the init container because it attempts to create a route that is already there. I'll need to add more logic to not create a route if it already exists.

**Comment by romfreiman** [comment]
@rwsu I would not remove the UI at this point - I think it will help us to debug. @filanov wdyt?

File: config/manifests/bases/assisted-service-operator.clusterserviceversion.yaml (hunk 19 above)
Comment (rwsu) [reply config/manifests/bases/assisted-service-operator.clusterserviceversion.yaml:16]: I will update this. Should NMStateConfig be included?

@filanov Currently the CSV will include and install all the CRDs in this directory: https://github.com/openshift/assisted-service/tree/master/internal/controller/config/crd/bases. Images and Cluster are still defined here as well as ClusterDeployment from hive. Will these 3 CRDs be eventually removed from this repo? Alternatively I can change the CSV to be selective and not include those.

File: cmd/main.go (hunk 9 above)
Comment (rwsu) [reply cmd/main.go]: The init container has replaced the in app code.

**Comment by filanov** [comment]
> > In addition we don't need to deploy UI, with hive and kube-api we expect the users not to use rest-api, deploying the UI may be confusing.
> 
> @filanov If the users will not use the rest-api, does this mean we don't need to create route for the service? Are the agents still communicating back using the rest-api?

We still need the route because hosts will connect to the service with rest-api.

File: config/assisted-service/postgres-secret.yaml (hunk 7 above)
Comment (rwsu) [reply config/assisted-service/postgres-secret.yaml]: Added an init container to randomly generate the password.

File: config/manifests/bases/assisted-service-operator.clusterserviceversion.yaml (hunk 19 above)
Comment (filanov) [reply config/manifests/bases/assisted-service-operator.clusterserviceversion.yaml]: Good question. in production clusterdeployment CRD should be deployed by hive. but for testing we probably need an option to deploy it as well.
InstallEnv and Agent will remain in this repo, we have a ticket to move them outside of internal dir. 
Regarding Cluster and Image we will do the cleanup this or next week.

File: docs/operator.md (hunk 12 above)
Comment (rwsu) [reply docs/operator.md:50]: I tested adding PVC to the operator bundle and it makes the operator undeployable. Operator becomes stuck at RequirementsNotMet status and olm-operator logs doesn't show what the problem could be.

File: bundle.Dockerfile (hunk 2 above)
Comment (rwsu) [reply bundle.Dockerfile:1]: bundle.Dockerfile is auto generated from "operator-sdk generate bundle -q --overwrite --version $(OPERATOR_VERSION)" in the operator-bundle target. I'll remove it from the repo, as it's an artifact from the generate bundle process.

File: config/rbac/controller_roles.yaml (hunk 17 above)
Comment (rwsu) [reply config/rbac/controller_roles.yaml]: This file was a copy of internal/controller/config/rbac/role.yaml. I've removed the temporary file and I've changed the name to assisted-service-manager-role in the original.

File: config/rbac/leader_election_role.yaml (hunk 18 above)
Comment (rwsu) [reply config/rbac/leader_election_role.yaml:5]: This file was a copy of internal/controller/config/rbac/kube_api_roles.yaml. I've removed the temporary file and I've changed the name to assisted-service-leader-election-role in the original.

File: PROJECT (hunk 4 above)
Comment (rwsu) [reply PROJECT]: I've corrected the repo url. Leaving the PROJECT file in the repo as it appears to be present in other operators' repos.

File: cmd/main.go (hunk 14 above)
Comment (rwsu) [reply cmd/main.go]: > Just making sure that we have permissions to edit Hive cluster deployment.

Yes, they are included in the CSV generated in the bundle. More specifically, the permissions in internal/controller/config/rbac/role.yaml are included in the CSV.

**Comment by rwsu** [comment]
> Just making sure that we have permissions to edit Hive cluster deployment.

Yes, they are included in the CSV generated in the bundle. More specifically, the permissions in internal/controller/config/rbac/role.yaml are included in the CSV.

**Comment by rwsu** [comment]
> @rwsu I would not remove the UI at this point - I think it will help us to debug. @filanov wdyt?

I've put the UI back in.

File: cmd/main.go (hunk 14 above)
Comment (rwsu) [reply cmd/main.go]: The code has been replaced with the initContainers which are namespace aware.

File: tools/wait_for_assisted_service.py [hunk 20]
```diff
@@ -25,6 +25,8 @@ def wait_for_request(url: str) -> bool:
 
 def main():
     deploy_options = handle_arguments()
+    if deploy_options.apply_manifest is False:
```
Comment (flaper87) [inline tools/wait_for_assisted_service.py:28]: Just a small nit:

```suggestion
    if not deploy_options.apply_manifest:
```

File: tools/deployment_options.py [hunk 21]
```diff
@@ -30,6 +30,12 @@ def load_deployment_options(parser=None):
         choices=['minikube', 'oc', 'oc-ingress', 'ocp'],
         default='minikube'
     )
+    parser.add_argument(
+        '--persistent-storage',
+        help='Persistent storage is used',
+        type=bool,
+        default=True
+    )
```
Comment (flaper87) [inline tools/deployment_options.py:38]:
```suggestion
    parser.add_argument(
        '--persistent-storage',
        help='Persistent storage is used',
        action=store_true
    )
```

I may have misunderstood what you are after here but, from what I can tell I think what you want is to use `--persistent-storage` and `--apply-manifest` as flags. In this case, using `store_true` as an action is better. Using `type` is normally not recommended, especially with the `bool` converter as it will convert empty strings to False and whatever value to True.

File: tools/deploy_ui.py [hunk 22]
```diff
@@ -35,13 +35,15 @@ def main():
...
     utils.check_output(cmd)
-    log.info("Deploying %s", dst_file)
-    utils.apply(
-        target=deploy_options.target,
-        namespace=deploy_options.namespace,
-        profile=deploy_options.profile,
-        file=dst_file
-    )
+
+    if deploy_options.apply_manifest is True:
```
Comment (flaper87) [inline tools/deploy_ui.py:39]:
```suggestion
    if deploy_options.apply_manifest:
```

nit :)

File: deploy/postgres/postgres-deployment.yaml [hunk 23]
```diff
@@ -36,7 +36,7 @@ spec:
                   name: assisted-installer-rds
                   key: db.password
           volumeMounts:
-            - mountPath: /var/lib/postgress
+            - mountPath: /var/lib/pgsql/data
```
Comment (flaper87) [inline deploy/postgres/postgres-deployment.yaml:39]: out of curiosity, is this change required? Wondering what it's related to.

**Review by flaper87 — COMMENTED** [review]
@rwsu good stuff! I still have to ramp up on everything and learn all the things. I only have nit comments for now. Hope you don't mind :D

File: tools/deployment_options.py (hunk 21 above)
Comment (rwsu) [reply tools/deployment_options.py:38]: They aren't meant to be flags.  You did point out a bug in how --persistent-storage was being used. The Makefile wasn't passing in a value like --persistent-storage $(PERSISTENT_STORAGE) in the Makefile to deploy_postgres.py, and that was probably the source of the confusion.

Good to know about action='store_true'. For me, It was counter-intuitive that the default value would be False.

File: deploy/postgres/postgres-deployment.yaml (hunk 23 above)
Comment (rwsu) [reply deploy/postgres/postgres-deployment.yaml:39]: /var/lib/postgress looked fishy to me. From the [docs] (https://hub.docker.com/_/postgres), the default should have been /var/lib/postgresql/data. When I actually looked at the postgres container logs, it was using /var/lib/pgsql/data. Looking at the docs more closely, I think we should also be setting a value for PGDATA.

Comment (flaper87) [reply deploy/postgres/postgres-deployment.yaml:39]: > Looking at the docs more closely, I think we should also be setting a value for PGDATA.

I think it's safer to set it, yes. I don't think this value will be changed upstream, which would break the current configs.


On that note, I wonder if we want to make this path configurable, instinct says we better don't do this yet unless there's a real need for it in the future.

File: tools/deployment_options.py (hunk 21 above)
Comment (flaper87) [reply tools/deployment_options.py]: > They aren't meant to be flags.

I'm a bit confused by this because they are currently being used as flags. The store true/false and they are used as booleans everywhere else. Maybe I'm missing something.

Could you elaborate on what cases these CLI params won't be used as flags? I'm sure there's a learning opportunity for me here :smile:

**Comment by rwsu** [comment]
Most of the feedback have been addressed and this PR is ready for final review. Replacing scality with the s3 filesystem implementation can be done in a follow up PR.

**Comment by filanov** [comment]
> @rwsu I would not remove the UI at this point - I think it will help us to debug. @filanov wdyt?

It will enforce us to work with CRs, and it's a good direction to not have any hacks.
We can add it back, but we need to have a ticket not to put it there when deploying to production.

File: config/rbac/leader_election_role_binding.yaml [hunk 24]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: leader-election-rolebinding
```
Comment (filanov) [inline config/rbac/leader_election_role_binding.yaml:4]: Maybe the name should be specific for assisted-service?

File: docs/operator.md [hunk 25]
```diff
@@ -0,0 +1,117 @@
+# Operator build and deployment
+
+## Prerequisites
+
+1. operator-sdk <https://sdk.operatorframework.io/docs/installation/>
```
Comment (filanov) [inline docs/operator.md:5]: if we are using the build container then it's not really needed right?

File: docs/operator.md [hunk 26]
```diff
@@ -0,0 +1,117 @@
...
+To generate the manifests and CSV in ./bundle, "make operator-bundle" first calls the "ocp-create-manifests" target. This target in turn calls "deploy-service-on-ocp-cluster and deploy-ui-on-ocp-cluster" while setting APPLY_MANIFESTS=False and APPLY_NAMESPACE=False. This causes the resource yamls to be created in ./build/assisted-installer/ but does not apply them.
+
+The relevant resource yamls are then copied to ./config/assisted-service where additional customizations are applied using Kustomize. The resulting yaml is then piped to operator-sdk
+creating the manifests and CSVs in ./bundle/manifests. 
+
+More information about bundles: <https://sdk.operatorframework.io/docs/olm-integration/generation/>
+
+### Create the bundle and index images
+
+```bash
```
Comment (filanov) [inline docs/operator.md:26]: We usually use skipper, is there a target that can be called from the makefile? 
I don't want to mix between podman and the docker that we use in the makefile. it will be confusing.
It's not a blocker for this PR just something to think about.

File: docs/operator.md [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+# Uncomment to set a mininum disk size allowed by the hardware validator.
+# By default the minimum disk size allowed is 120GB. The patch sets the
+# minimum disk size to 20GB.
+#- assisted-service-configmap-patch-hw-validator-min-disk-size.yaml
+```
+
+```
+# Uncomment to use a custom assisted-service image in the deployment
+#- assisted-service-patch-image.yaml
+```
```
Comment (jparrill) [inline docs/operator.md:117]: Hey @rwsu I would add some Sample CR's in order to let the users try it out, at least the ClusterDeployment and the InstallEnv ones. This way you will help them to test it faster.

**Review by filanov — APPROVED** [review]

File: docs/operator.md (hunk 27 above)
Comment (filanov) [reply docs/operator.md:117]: +1 
Can probably done in a different PR

File: Makefile [hunk 28]
```diff
@@ -353,3 +378,35 @@ delete-minikube-profile:
...
+	cp ./build/assisted-installer/deploy_ui.yaml config/assisted-service
+	#operator-sdk generate kustomize manifests -q
+	#cd config/manager && $(KUSTOMIZE) edit set image controller=$(IMG)
+	kustomize build config/manifests | operator-sdk generate bundle -q --overwrite --version $(OPERATOR_VERSION) $(BUNDLE_METADATA_OPTS)
+	operator-sdk bundle validate ./bundle
+
+# Build the bundle image.
+.PHONY: operator-bundle-build
+operator-bundle-build:
+	podman build -f bundle.Dockerfile -t $(BUNDLE_IMG) .
```
Comment (djzager) [inline Makefile:412]: I don't see the `bundle.Dockerfile` anywhere. Is it generated somewhere or just missing?

Comment (rwsu) [reply Makefile:412]: I removed it from the repo and added to .gitignore. It is generated by "make operator-build".

File: deploy/postgres/postgres-deployment.yaml (hunk 23 above)
Comment (rwsu) [reply deploy/postgres/postgres-deployment.yaml:39]: One of the QE folks is having problems deploying postgres using a NFS backed PV:

`11:39:11.715 UTC [72] FATAL:  data directory "/var/lib/pgsql/data/userdata" has wrong ownership`

Seems to point to setting PGDATA. Will need further investigation.
